### ISSUE TITLE
Quantized Centernet model produces defected results when run on Android via NDK

### ISSUE THREAD

**Issue by maxfiedler** [pytorch/pytorch#47295]
## 🐛 Bug

We have a CenterNet implementation traced as a TorchScript model and saved as both in a quantized and unquantized variant.

One of the results our CenterNet modle produces is a heatmap that contains for every pixel a score corresponding to the likelihood of the pixel being the center of an object visible in the input image. 

A decent example of such a centerpoint heatmap could look like this:

![image](https://user-images.githubusercontent.com/4192637/98020314-24703280-1e03-11eb-8c1c-75f01a802d4b.png)

- When we run the forward pass in different Android Apps directly via the Java bindings, both variants produces decent results as the one seen above. 
- When we run the forward pass via the C++ front-end directly on a Ubuntu 18.04 docker container on our x86 PCs both variants produces decent results as the one seen above
- When we run the forward pass on Android Apps via binding a C++ frontend into the app through NDK and JNI-calls , we see that the unquantized model gives the decent result we expect, but the quantized model produces a centerpoint heatmap with clear artifacts. An example can be seen below:

![image](https://user-images.githubusercontent.com/4192637/98020796-c4c65700-1e03-11eb-8fab-14b2b6f2d775.png)

As you can see it shows a very particular pattern that is consistently produced on every run in different apps while capturing different scenery. 

We also used both a pre-compiled version of libtorch following this tutorial https://pytorch.org/tutorials/recipes/android_native_app_with_custom_op.html

as well as compiling libtorch from source, producing these artifacts in both cases. 

## To Reproduce

Unfortunately, we cannot share our model. But we might be able to provide a minimal app with simple model soon to reproduce

## Environment

 - PyTorch Version (e.g., 1.0): we tried with 1.6 and 1.5
 - OS (e.g., Linux): Ubuntu 18.04 docker container for execution on x86, Android 10.0 for mobile
 - How you installed PyTorch (`conda`, `pip`, source): both from source and via Android build.gradle ('com.android.tools.build:gradle:4.0.1')
 - Build command you used (if compiling from source): will provide later
 - Python version: 3.7
 - C++ version: C++17
 - cmake version: 3.17
 - CUDA/cuDNN version: not using CUDA
 - GPU models and configuration: running on CPU only
 - Any other relevant information:

## Additional context

This is probably for now a corner case application, but it would be really great to be able to develop re-usable libraries and application code with the C++ frontend for deployment to both Mobile and other ARM based platforms. 

From your experience, is the integration of libtorch into iOS less shaky than into Android?

cc @jerryzh168 @jianyuh @dzhulgakov @raghuramank100 @jamesr66a @vkuzo

**Comment by vkuzo** [comment]
The bug fix in https://github.com/pytorch/pytorch/pull/46077 could be related.  If you were using hardsigmoid, tanh, sigmoid, the layer inputs were incorrect in QNNPACK in some conditions when using NHWC memory format.  The visual artifacts we saw in affected models were similar to your image above.  This was fixed in v1.7, could you check if v1.7 fixes your issue?

**Comment by mmagst** [comment]
Unfortunately, in this case we are using simple ReLUs everywhere and use NCHW. But I agree that this does look a lot like some memory layout issue. We do have a few interpolation calls in the model, could these be causing some trouble here?

We also unfortunately were able to verify that 1.7 did not fix the issue.

**Comment by maxfiedler** [comment]
@vkuzo  thanks for getting back to us. Btw., we also encountered another bug that might have a related root cause, see https://github.com/pytorch/pytorch/issues/46995

**Comment by maxfiedler** [comment]
Also, we will soon try to run the same model (quantized and unquantized) in a Linux based ARM64 emulator in order to pinpoint if the platform is the Issue or running on Android. We will let you know what we discover here.

**Comment by soandrey** [comment]
I'm facing the same issue but not on Android, I'm running libtorch for C++ on x86, and the issue appeared after moving from libtorch 1.6 to libtorch 1.7.

The non quantized model still gives me the expected output:

![image](https://user-images.githubusercontent.com/19828234/98257283-c9177f00-1f7f-11eb-96b8-7b20848f7d26.png)


But the quantized version of that model has the same pattern as the one above:

![image](https://user-images.githubusercontent.com/19828234/98257399-efd5b580-1f7f-11eb-8996-ba67d47a82dd.png)

**Comment by maxfiedler** [comment]
And to add, this happens with libtorch 1.7 both in case of a model scripted with PT1.6 and one with PT1.7

**Comment by maxfiedler** [comment]
so we have a few more updates on this issue. 
We experimented with our MobileNet backbone standalone, feeding the unquantized and quantized model torch::random and torch::ones inputs and visualizing the output featuremaps in the following videos.

The videos work as follows: First 16 frames are the 16 final feature maps of the MobileNet backbone, the second 16 frames are then the 16 feature maps of the quantized MobileNet backbone. 

The feature maps of the unquantized model look pretty much like what you would expect, the feature maps produced by the quantized one show distinctive stripe patterns that also only slightly shift from feature map to feature map. 

This raises the suspicion that the underlying memory is somehow accessed in the wrong way by the resulting output Tensors of the layers. 

You can find the videos here: 
https://www.dropbox.com/sh/4l7bkmfe8bhzd72/AAAlPmX4VyGgzRxbI0oU8uvTa?dl=0

The last video is produced from the outputs of a model that contains only a single convolution layer

**Comment by dreiss** [comment]
Did I read the original report correctly that if you run on Android using the Java bindings (`org.pytorch.Module`, etc.), you get the correct output, but running the *same model on the same device* with the raw C++ API (`torch::jit::load`, etc.), you get the wrong output?  If that is correct, my guess would be that the problem is in the processing of the output.  Maybe it's returning a channels-last output, and you're not accounting for the strides?  The code in the Java bindings that handles this is https://github.com/pytorch/pytorch/blob/master/android/pytorch_android/src/main/cpp/pytorch_jni_common.cpp#L192 .  Are you able to share the C++ code you're using to render the problematic frames?

It's possible that our QNNPACK bindings are changing the memory format silently.  Does this reproduce on Ubuntu if you set the quantized engine to QNNPACK (like https://github.com/pytorch/pytorch/blob/master/android/pytorch_android/src/main/cpp/pytorch_jni_jit.cpp#L108 or https://pytorch.org/docs/stable/quantization.html).

**Comment by maxfiedler** [comment]
Hi @dreiss 
thanks for getting back to us. Yes, the same model worked without any issues using the Java bindings both quantized and unquantized. 
We will try to share a minimal app environment with you in which you can run a simple model both via Android and as a standalone C++ build on Monday next week. 

Regarding your suggestion for QNNPACK we will try this out in case we were not doing this yet. 
Can you comment on this @mmagst @soandrey

**Comment by dreiss** [comment]
> Regarding your suggestion for QNNPACK we will try this out in case we were not doing this yet.

To be clear, I'm interested if enabling this on *desktop or server* causes the bug to appear.  Setting this should have no effect on Android, since QNNPACK is (currently) the only quantized backend we have available on Android.

**Comment by mmagst** [comment]
Hi @dreiss ,

Thanks a lot for your insightful reply, it seems that your conjecture is right and the memory format is changed silently by the QNNPACK operations. We have quantized the following toy model with the QNNPACK engine:

`
class ConvModel(nn.Module):

    def __init__(self):
        super(ConvModel, self).__init__()
        self.conv = nn.Conv2d(3, 6, kernel_size=3)
        self.quant = QuantStub()
        self.dequant = DeQuantStub()

    def forward(self, x):
        x = self.quant(x)
        x = self.conv(x)
        x = self.dequant(x)
        return x
`

This does in fact cause a similar artifact in the output as shown in the original post _on a Ubuntu workstation using the C++ API_ which again looks very strongly like a misalignment of the output's memory layout and its processing.

And indeed, returning `x.contiguous()` resolves the problem. We can also confirm that the further processing of the output assumes NCHW format. We will provide a more detailed answer soon, but it really does seem like a silent switch from NCHW to NHWC is the issue here.

Coming from deploying for Java bindings, this was a sneaky one. I guess we could have read somewhere about this change in memory format but it really was not obvious to us. As a suggestion, it might be helpful for developers to mention this potential pitfall somewhere more prominent in the docs.

Thanks for helping us out here!

**Comment by dreiss** [comment]
This is a bug.  All of our convolutions should be preserving the memory format.

**Comment by kimishpatel** [comment]
Looks like a bug in qconv.

**Comment by mmagst** [comment]
Agree, if you expect the convolutions to always preserve the memory format, then this clearly is a bug in the qconv. 

So, the automatic checks and conversions for the memory format in the Java bindings you pointed to above were not put there to account for the change in memory layout then?

Please find below the toy example I used to reproduce the bug. I used Pytorch 1.7 stable for this. Are there any other specifics you would like to receive from our end?

```
import torch
from torch.quantization import DeQuantStub
from torch.quantization import QuantStub


class ConvModel(torch.nn.Module):

    def __init__(self):
        super(ConvModel, self).__init__()
        self.conv = torch.nn.Conv2d(3, 6, kernel_size=3)
        self.quant = QuantStub()
        self.dequant = DeQuantStub()

    def forward(self, x):
        x = self.quant(x)
        x = self.conv(x)
        x = self.dequant(x)  # .contiguous()
        return x


def main():
    model = ConvModel()

    backend = 'qnnpack'
    qconfig = torch.quantization.get_default_qconfig(backend)
    model.qconfig = qconfig
    torch.backends.quantized.engine = backend

    quant_model = torch.quantization.prepare(model, inplace=False)

    for i in range(100):
        print(f'quant iter {i}')
        _ = quant_model(torch.rand(1, 3, 224, 224))

    torch.quantization.convert(quant_model, inplace=True)
    
    torch.jit.save(torch.jit.script(model), 'conv.pt')
    torch.jit.save(torch.jit.script(quant_model), 'conv_quant.pt')


if __name__ == '__main__':
    main()

**Comment by mmagst** [comment]
Here is a C++ snippet showing how we access the output tensor. This will clearly fail if the tensor is not in NCHW and likely then produce a pattern such as the one shown above.

```
#include <torch/script.h>
#include <opencv2/core.hpp>
#include <vector>
#include <iostream>


std::vector<cv::Mat> tensor_to_vector_of_mat_per_channel(torch::Tensor &input, int mat_type) {
    std::vector<cv::Mat> mat_vector;
    torch::IntArrayRef shape = input.sizes();
    size_t ndim = shape.size();
    // the standard encoding for image Tensors in Pytorch is NCHW [batch_size, channels, height, width]
    // with the batch dimension being optional
    if (ndim == 4) {
        // if there is a batch dimension, remove it
        input = torch::squeeze(input, 0);
    }
    int num_channels = shape.at(ndim - 3);
    for (int i = 0; i < num_channels; ++i) {
        torch::Tensor sub_tensor = input.index({i, torch::indexing::Ellipsis});
        mat_vector.emplace_back(tensor_to_mat_single_channel(sub_tensor, mat_type));
    }
    return mat_vector;
}

cv::Mat tensor_to_mat_single_channel(torch::Tensor tensor, int mat_type) {
    tensor = tensor.squeeze();
    int rows, cols;
    if (tensor.dim() == 2) {
        rows = tensor.sizes()[0];
        cols = tensor.sizes()[1];
    } else if (tensor.dim() == 1) {
        rows = 1;
        cols = tensor.sizes()[0];
    } else {
        std::cerr << "TENSOR_TO_MAT_SINGLE_CHANNEL: Dimensions have to be two or one!" << std::endl;
        return cv::Mat::zeros(0, 0, CV_32FC1);
    }

    cv::Mat mat;
    try {
        tensor = tensor.to(torch::kF32);
        mat = cv::Mat(rows, cols, CV_32FC1, tensor.data_ptr());
        mat.convertTo(mat, mat_type);
    } catch (const c10::Error& e) {
        std::cout << "an error has occured : " << e.msg() << std::endl;
    }
    return mat;
}

void run_model() {
    auto module = torch::jit::load("path_to_model");
    auto tensor = torch::rand({1, 15, 320, 320});
    auto inputs = std::vector<torch::jit::IValue>{tensor};

    auto model_output = module.forward(inputs).toTensorList();

    // Depending on output use either
    auto channels = tensor_to_vector_of_mat_per_channel(model_output, CV_32F);
    auto channel = tensor_to_mat_single_channel(model_output, CV_32FC1);
}

**Comment by kimishpatel** [comment]
Thanks @mmagst for the code snippet. Will put up a fix soon.

**Comment by dreiss** [comment]
> So, the automatic checks and conversions for the memory format in the Java bindings you pointed to above were not put there to account for the change in memory layout then?

No.  While our conv ops are always expected to preserve memory layout, not all operations will return contiguous tensors.  For example, a slice operation will almost always return a non-contiguous tensor.  Because our Java bindings do not yet support non-contiguous, we must call `.contiguous()` on all tensors before passing them to Java.  However, your C++ code gets the real PyTorch tensor, which is not contiguous.

Calling `.contiguous()` on your output tensor (either in the model, or outside of it) is the correct short-term workaround for this issue until we fix the bug.  In the long term, keep in mind that other models might return non-contiguous tensors even if there is no bug (for example, if they end with a slice operation).  Therefore, I would suggest that you either (a) explicitly document within your system that all models are expected to return contiguous tensors, (b) always call `.contiguous()` on the tensor you get back from the model (which is an O(1) operation if the tensor is already contiguous), or (c) use a stride-aware mechanism (like TensorAccessor) to extract data from the tensor.

**Comment by mmagst** [comment]
Thanks for the clarification. Then we'll stick with this workaround for the moment and will generally keep the memory format in mind when converting our output tensors. Thanks again for the help!

**Comment by supriyar** [comment]
> Thanks @mmagst for the code snippet. Will put up a fix soon.

@kimishpatel did you get a chance to work on a fix for this?

**Comment by axitkhurana** [comment]
@supriyar I'm looking into this

**Comment by kimishpatel** [comment]
> > Thanks @mmagst for the code snippet. Will put up a fix soon.
> 
> @kimishpatel did you get a chance to work on a fix for this?

@axitkhurana is on it. Akshit, can you give any estimate on this?

**Comment by axitkhurana** [comment]
Sure, will try to get it in this week

**Comment by mmagst** [comment]
Thanks, looking forward to using the fixed version!
